python/MooseDocs/extensions/content.py

```diff
@@ -281,3 +338,63 @@ def createHTML(self, parent, token, page):
 
     def createLatex(self, parent, token, page):
         return None
+
+class RenderContentOutline(components.RenderComponent):
+    def createHTML(self, parent, token, page):
+        if token['location'] is not None:
+            func = lambda p: p.local.startswith(token['location']) and isinstance(p, pages.Source)
+            nodes = self.translator.findPages(func)
+        elif token['pages']: # should I even check? It would have to be pages option at this point
+            nodes = [self.translator.findPage(p) for p in token['pages']]
+
+        ul = html.Tag(parent, 'ul', class_='moose-outline')
+        for node in nodes:
+            pageref = str(node.relativeDestination(page))
+
+            # this loop produces a KeyError: 'headings' when it re-renders after changing
+            # a markdown file while already serving, but not on the first build --serve.
+            # can we fix this?
+            for k, v in node['headings'].items():
+                if v['level'] in token['levels'] and k not in token['hide']:
+                    classref = 'moose-outline-{}'.format(v['level'])
+
+                    if v['level'] is 1:
+                        ol1 = html.Tag(ul, 'ol', class_=classref)
+                        li = html.Tag(ol1, 'li')
+                    elif v['level'] is 2:
+                        ol2 = html.Tag(ol1, 'ol', class_=classref)
+                        li = html.Tag(ol2, 'li')
+                    elif v['level'] is 3:
+                        ol3 = html.Tag(ol2, 'ol', class_=classref)
+                        li = html.Tag(ol3, 'li')
+                    elif v['level'] is 4:
+                        ol4 = html.Tag(ol3, 'ol', class_=classref)
+                        li = html.Tag(ol4, 'li')
+                    elif v['level'] is 5:
+                        ol5 = html.Tag(ol4, 'ol', class_=classref)
+                        li = html.Tag(ol5, 'li')
+                    elif v['level'] is 6:
+                        ol6 = html.Tag(ol5, 'ol', class_=classref)
+                        li = html.Tag(ol6, 'li')
+                    # else: error? is this even possible?
+
+                    url = pageref + '#{}'.format(k)
+                    head = heading.find_heading(node, k)
+                    link = core.Link(None, url=url)
+                    head.copyToToken(link)
+                    self.renderer.render(li, link, page)
```

Comment:
There is a better way to do this, but I am not sure what that is and will need to play with this a bit. I will do my best to get to this today,